### PR TITLE
polish(davide) add example of Gail entry + config for Mujoco and Cartpole

### PR DESCRIPTION
**Pull request by davide97l**
## Description
This repo adds a config + entry to train an imitation policy to imitate an expert policy using GAIL. However the same entry can be generalized to any reward model whose interface is similar to GAIL. It includes the data collection phase and can save both the IL policy and reward model during training.

Important: this PR also changes the formula of the loss and reward of GAIL, changes have been approved after a discussion with @Will-Nie and @PaParaZz1. Now the loss function is according to the formula defined in the original paper.
Check the related comment for more details.
 
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

**Comment by davide97l** [comment]
As we can see from the plot above, with the new formula GAIL can train Cartpole faster:
aaa

The loss of GAIl changes in the following way:
photo_2021-11-02_16-00-56
.

**Comment by davide97l** [comment]
image
The reward function has also been changed, this further improves GAIL performance.

File: ding/reward_model/gail_irl_model.py [hunk 1]
```diff
@@ -166,10 +166,10 @@ def estimate(self, data: list) -> None:
         res = concat_state_action_pairs(data)
         res = torch.stack(res).to(self.device)
         with torch.no_grad():
-            reward = -self.reward_model(res).squeeze(-1).cpu()
+            reward = self.reward_model(res).squeeze(-1).cpu()
         reward = torch.chunk(reward, reward.shape[0], dim=0)
         for item, rew in zip(data, reward):
-            item['reward'] = rew
+            item['reward'] = -torch.log(rew)
```
Comment (Will-Nie) [inline ding/reward_model/gail_irl_model.py:172]: shall we add a small number in log to avoid potential explosion? e.g. -torch.log(rew + 1e-8)

Comment (davide97l) [reply ding/reward_model/gail_irl_model.py:172]: It makes sense, I will add it, thank you for your finding.

File: dizoo/classic_control/cartpole/config/cartpole_dqn_gail_config.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ),
+    reward_model=dict(
+        type='gail',
+        input_size=5,
+        hidden_size=64,
+        batch_size=64,
+        learning_rate=1e-3,
+        update_per_collect=100,
+        expert_data_path='cartpole_dqn/expert_data_train.pkl',
+        load_path='cartpole_dqn_gail/reward_model/ckpt/ckpt_last.pth.tar',
```
Comment (Will-Nie) [inline dizoo/classic_control/cartpole/config/cartpole_dqn_gail_config.py:19]: should here be ckpt_best?  Also, what is 'load_path' here for?

Comment (Will-Nie) [inline dizoo/classic_control/cartpole/config/cartpole_dqn_gail_config.py:19]: Is load_path refering to the well-trained reward model you finally save? can we make this clearer? (let other users directly understanding that 'load_path' refers to the final reward model the algorithm learned)

Also, Why do we need load_path key here?

File: dizoo/classic_control/cartpole/entry/cartpole_dqn_gail_main.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    # Learner's after_run hook.
+    learner.call_hook('after_run')
+    # save reward model
+    path = os.path.join(cfg.exp_name, 'reward_model', 'ckpt')
+    if not os.path.exists(path):
+        try:
+            os.makedirs(path)
+        except FileExistsError:
+            pass
+    path = os.path.join(path, 'ckpt_last.pth.tar')
```
Comment (Will-Nie) [inline dizoo/classic_control/cartpole/entry/cartpole_dqn_gail_main.py:149]: maybe 'ckpt_final_reward_model.pth.tar' is more clearer? ( as we have ckpt_best.pth.tar before, referring to the best model for the training the policy.)

File: dizoo/classic_control/cartpole/config/cartpole_dqn_gail_config.py (hunk 2 above)
Comment (davide97l) [reply dizoo/classic_control/cartpole/config/cartpole_dqn_gail_config.py:19]: There are 2 load_path:
- policy.load_path is the path where the state_dict of the **policy** is saved
- reward_model.load_path is the path where the state_dict of the **reward model** is saved

I will write a comment in the config file to clarify it

File: dizoo/classic_control/cartpole/entry/cartpole_dqn_gail_main.py (hunk 3 above)
Comment (davide97l) [reply dizoo/classic_control/cartpole/entry/cartpole_dqn_gail_main.py:149]: Line 143: `path = os.path.join(cfg.exp_name, 'reward_model', 'ckpt')` creates a new path (`reward_model/ckpt`) where to save the state dict of the reward model. Given the explicit name, it should be clear that are pth.tar files inside that directory refer to the state dict of the reward model.

File: dizoo/classic_control/cartpole/config/cartpole_dqn_config.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
         replay_path='cartpole_dqn/video',
     ),
     policy=dict(
-        load_path='',
+        load_path='cartpole_dqn/ckpt/ckpt_best.pth.tar',
```
Comment (PaParaZz1) [inline dizoo/classic_control/cartpole/config/cartpole_dqn_config.py:13]: remove related modifications like this for compatibility

File: dizoo/box2d/lunarlander/entry/lunarlander_dqn_gail_main.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
+from typing import Optional, Tuple
```
Comment (PaParaZz1) [inline dizoo/box2d/lunarlander/entry/lunarlander_dqn_gail_main.py:1]: can you merge atari/cartpole/lunarlander main to a general gail_main?

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/reward_model/gail_irl_model.py [hunk 6]
```diff
@@ -119,7 +163,7 @@ def _train(self, train_data: torch.Tensor, expert_data: torch.Tensor) -> float:
         loss_1: torch.Tensor = torch.log(out_1 + 1e-8).mean()
         out_2: torch.Tensor = self.reward_model(expert_data)
         loss_2: torch.Tensor = torch.log(1 - out_2 + 1e-8).mean()
-        loss: torch.Tensor = loss_1 + loss_2
+        loss: torch.Tensor = - (loss_1 + loss_2)
```
[the remark this answers is not in the document]
Comment (davide97l) [reply ding/reward_model/gail_irl_model.py:166]: I will add this comment: log(x) with 0<x<1 is negative, so to reduce this loss we have to minimize the opposite

**Review by PaParaZz1 — APPROVED** [review]
